### ISSUE TITLE
Using the probability of each token in the vocabulary in training

### ISSUE THREAD

**Issue by samialabed** [samialabed/method-name-prediction#5]
Branch: [Probability-CNN](https://github.com/samialabed/method-name-prediction/tree/probability-cnn/src/models)

Currently, [CNN Attention](https://github.com/samialabed/method-name-prediction/blob/probability-cnn/src/models/cnn_attention.py) returns the probability of each subtoken in the vocabulary associated with its probability, as per the paper. 
Thus the output shape is [batch size, size of the vocabulary]

However, this isn't suitable for training - the training labels are of the shape [word length (padded to a maximum of max_chunk_size), embedding dim)

@mallamanis Can you please  clarify how the vocabulary probabilities resulting from CNN on a sequence of token body C is then used to predict the sequence M (the method name).

I am using sparse_cross_entropy as a loss function - can be seen (here)[https://github.com/samialabed/method-name-prediction/blob/probability-cnn/notebooks/02-masking-exploration.ipynb]

**Comment by samialabed** [comment]
Reference to the related code in the paper:
![image](https://user-images.githubusercontent.com/6316945/53805095-2a808280-3f41-11e9-971a-c5a70f7f8847.png)

Shouldn't the output be of shape [batch size, max chunk size, vocabulary size]?

**Comment by mallamanis** [comment]
I am not sure that I fully understand you question but let me give it a try:

This is the "normal" (not copying) attention. At each timestep, the network predicts the next subtoken of the name, thus at each timestep the output is `[batch size, size of vocabulary]`, ie for each element in the batch a probability distribution over the subtokens to generate at the current step. Now, if you unroll this in time, you will get a `[batch size, max output length, size of vocabulary]`, ie one for each timestep.

Essentially, the code you see in the paper is about one timestep. 

Also, the training labels you have (with padding) are [batch size, max output length], which is what [`sparse_softmax_cross_entropy`](https://www.tensorflow.org/api_docs/python/tf/losses/sparse_softmax_cross_entropy) actually takes.

**Comment by samialabed** [comment]
Thank you! This makes so much sense, yes this is for the "normal" attention.
I have updated the code in https://github.com/samialabed/method-name-prediction/commit/cf7b30c02aa1128799792ccf54922bb97b504373 to take timesteps (each subtoken being a timestep) in the layers so it can predict the whole sequence. 

However, the model still outputs <s></s>pading-padding-padding and nothing else, I can't seem to force it to overfit even on a small dataset, the [notebook here](https://github.com/samialabed/method-name-prediction/blob/probability-cnn/notebooks/02-masking-exploration.ipynb) can show the way I am training the model. 

I understand I want to use beam search to generate reasonable values, but even when I look at top_5 possible values returned by the model they don't make any sense in relation to the input. 

Do you think I should use my own custom loss function and remove the padding from the y_true? or sparse_softmax already take care of that

**Comment by mallamanis** [comment]
As I've never worked with Keras, it's hard to understand the code.

* I am not sure that [this](https://github.com/samialabed/method-name-prediction/blob/638e2084d4731398d9283b9416dea024366a4f15/src/models/cnn_attention.py#L74) is correct. You are doing a dot product with the embeddings of the input and not the embedding matrix. What you want is to do the dot product with `E` which should be internal to `embedding_layer` (and not `token_embeddings`).
* The bias term `b` is missing.
* I don't think that the loss function is the problem. 
* When trying to overfit, remove dropout.
* Forget about beam search at this moment. When overfitting the greedy decoding should suffice.

**Comment by samialabed** [comment]
Thank you! Addressed in https://github.com/samialabed/method-name-prediction/commit/36e641bd9cfcc086d553e019aa6ce2305a17ada2 

This works well now
